kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineEventStore.kt

```diff
@@ -227,45 +227,66 @@ class AppEngineEventStore(private val kind: String = "Event", private val messag
         return GetEventsResponse.Success(aggregates)
     }
 
-    override fun getEvents(aggregateId: String, aggregateType: String): GetEventsResponse {
-        val dataStore = DatastoreServiceFactory.getDatastoreService()
+    override fun getEvents(aggregateId: String, aggregateType: String, index: Long?): GetEventsResponse {
+      val dataStore = DatastoreServiceFactory.getDatastoreService()
 
+      if (index == null) {
         val snapshotKey = snapshotKey(aggregateId, aggregateType)
         val snapshotEntity: Entity?
 
         snapshotEntity = try {
-            dataStore.get(snapshotKey)
+          dataStore.get(snapshotKey)
         } catch (ex: EntityNotFoundException) {
-            null
+          null
         }
 
         val aggregateIndex = snapshotEntity?.getProperty("aggregateIndex") as Long? ?: 0
```

Comment:
I suppose only this line could be moved in the if condition as it's used for retrieving of the index.  If the user specifies the index then a specific one will be loaded, otherwise the index will be retrieved and concrete value will be used.

Another option would be the client code to specify whether retrieveAllEvents option to be used which will perform.

0..aggregateIndex => load all entities and events ..

Reply:
Extracted duplicated logic.
I don't want to allow the response to contains all events because can be too big.

Reply:
Yes, but now the client has to provide aggregateIndex externally and it's not clear for me how he is able to retrieve it.

Reply:
Common case where this index will be needed is restore of some view so loading of events start from first event(index 0) and end when not found aggregates events with current index.

This index will be used as mechanism for sequence loading of event on portions for single aggregate.

Reply:
This requires one additional roundtrip which doesn't sound like a good design.